static_frame/core/index_hierarchy.py

```diff
@@ -1123,154 +1418,405 @@ def relabel_at_depth(self,
+        for level in depth_level:
+            index = self._indices[level]
 
-        def gen() -> tp.Iterator[np.ndarray]:
-            for depth_idx in range(self.depth):
-                if depth_idx in target_depths:
-                    yield values
+            new_index: tp.Dict[tp.Hashable, int] = {}
+            index_remap: tp.Dict[tp.Hashable, tp.Hashable] = {}
+
+            for label_idx, label in enumerate(index.values):
+                new_label = get_new_label(label)
+
+                if new_label not in new_index:
+                    new_index[new_label] = len(new_index)
                 else:
-                    yield self._blocks._extract_array_column(depth_idx)
+                    index_remap[label_idx] = new_index[new_label]
 
-        return self.__class__._from_type_blocks(
-                TypeBlocks.from_blocks(gen()),
+            new_indices[level] = index.__class__(new_index)
+
+            indexer = np.array([index_remap.get(i, i) for i in self._indexers[level]])
+            indexer.flags.writeable = False
+            new_indexers[level] = indexer
+
+        return self.__class__(
+                indices=new_indices,
+                indexers=new_indexers,
                 name=self._name,
-                index_constructors=tuple(self._levels.index_types()),
-                own_blocks=True
                 )
 
     def rehierarch(self: IH,
-            depth_map: tp.Sequence[int]
+            depth_map: tp.Sequence[int],
             ) -> IH:
         '''
         Return a new :obj:`IndexHierarchy` that conforms to the new depth assignments given be `depth_map`.
         '''
-        if self._recache:
-            self._update_array_cache()
-
-        index_constructors = tuple(self._levels.index_types())
-
-        index, _ = rehierarch_from_type_blocks(
+        rehierarched_blocks, _ = rehierarch_from_type_blocks(
                 labels=self._blocks,
-                index_cls=self.__class__,
-                index_constructors=index_constructors,
                 depth_map=depth_map,
                 )
-        return index #type: ignore
+        return self.__class__._from_type_blocks(
+            blocks=rehierarched_blocks,
+            index_constructors=self._index_constructors,
+            own_blocks=True,
+            )
 
-    #---------------------------------------------------------------------------
+    def _get_outer_index_labels_in_order_they_appear(self: IH) -> tp.Sequence[tp.Hashable]:
+        '''
+        Index could be [A, B, C]
+        Indexers could be [2, 0, 0, 2, 1]
 
-    def _loc_to_iloc(self,
-            key: tp.Union[GetItemKeyType, HLoc]
-            ) -> GetItemKeyType:
+        This function return [C, A, B] -- shoutout to my initials
+        '''
+        # get the outer level, or just the unique frame labels needed
+        labels = self.values_at_depth(0)
+        label_indexes = sorted(ufunc_unique1d_positions(labels)[0])
```

Comment:
Can we use a NumPy sort here? Does sort stability matter?

Reply:
You mean in a sorted order, not the order they appear? As `ufunc_unique1d_positions` hands back arrays, however, I still do not see why we need to create a Python list and sort it, rather than just sorting the NumPy array.

Reply:
Oh sorry - I misread your original comment. Yes I will use `np.sort`!